### ISSUE TITLE
t-rdfMerger doesn't merge named graphs

### ISSUE THREAD

**Issue by jindrichmynarz** [UnifiedViews/Plugins#269]
t-rdfMerger doesn't merge named graphs. For example, I have a pipeline with 2 instances of t-filesToRdf that send RDFDataUnits into t-rdfMerger. URIs of named graphs in these RDFDataUnits are set using the fixed symbolic name option in t-filesToRdf as `http://data.sazp.sk/vocab/contaminated-sites` and `http://purl.org/procurement/public-contracts`. These RDFDataUnits are fed into t-rdfMerger, which passes the merged data into l-rdfToVirtuoso, which is set to work in the *Per graph mode*. When the pipeline is run, instead of loading the data into a single graph, 2 separate graphs with auto-generated URIs (`http://unifiedviews.eu/resource/internal/dataunit/exec/503/dpu/619/du/1/entry/1` and `http://unifiedviews.eu/resource/internal/dataunit/exec/503/dpu/620/du/1/entry/1`) are created.

There are several problems:

1. Data from the RDFDataUnits is not merged.
2. t-rdfMerger outputs multiple named graphs instead of a single one.
3. t-rdfMerger doesn't respect fixed symbolic names as named graph URIs and auto-generates new ones.

@tomas-knap: @mkl-eea-sk indicated that this is @skodape's DPU. Please see if Petr can work on this and reassign.

Note that #268 should be taken into account when resolving this issue with regards to the approach to setting named graph URIs.

Context: Tested on version 2.0.1 of the DPU and the most recent develop version (7dc113898b4c9192cc8d5cd8647b6768b0e98de5).

**Comment by skodape** [comment]
@jindrichmynarz htps://github.com/mff-uk/DPUs/tree/master/dpu/t-graphMerger does merge on graph level, perhaps you can give it a try?

**Comment by jindrichmynarz** [comment]
Indeed, t-graphMerger (compiled from the [most recent commit](https://github.com/mff-uk/DPUs/commit/45a72fea31a05c879c3e21ef4b20eea5871a95bf)) produces only a single graph with an auto-generated URI containing the merged data. Can we fix t-rdfMerger according to t-graphMerger?

**Comment by tomas-knap** [comment]
Solution: Add checkbox to t-rdfMerger "Physically merge graphs into one graph" (by default unchecked)  - and use the implementation from t-graphMerger to realize physical merge of graphs to one graph (see: htps://github.com/mff-uk/DPUs/tree/master/dpu/t-graphMerger). 

Estimation: 2h including testing

**Comment by jindrichmynarz** [comment]
That would fix the problem of unmerged named graphs URIs. However, the most recent version of the DPU doesn't even merge the RDF data itself.

This may be my misunderstanding of the intended function of this DPU. Is it intended to be used only to merge multiple RDFDataUnits (each containing 1+ named graphs) to a single RDFDataUnit (containing all named graphs from the merged RDFDataUnit)? If that is the case, shouldn't we also have another DPU (or option in this DPU) that actually merges data?

**Comment by skrchnavy** [comment]
@mkl-eea-sk what was behaviour in UV 1.X before moving from Core to Plugins?

**Comment by mkl-eea-sk** [comment]
same, but the current code seems to do merging into one symbolic name. so if this code is not doing this, it is bug in it. its simple, browse the code - only one addNewDataGraph is called on output data unit... damm, how can it produce two graphs...

----- Original Message -----

**Comment by tomas-knap** [comment]
@jindrichmynarz Instead of having two DPU rdfMerger a rdfPhysicalMerger, which may be confusing for the users, I suggest to have just one rdfMerger working as is intended now (just grouping all input graph under one data unit) and also being able to work in mode where it physically copies all input data to one graph. 

So to solve this issue: Check the existing functionality of grouping input graphs under one symbolic name, whether it works correctly. Add functionality of physical merger (as an option - see above), inspiration: htps://github.com/mff-uk/DPUs/tree/master/dpu/t-graphMerger

**Comment by mkl-eea-sk** [comment]
See this code:
```
  private static final String QUERY_COPY = "INSERT { ?s ?p ?o } WHERE { ?s ?p ?o }";

    @DataUnit.AsInput(name = "rdfInput")
    public RDFDataUnit input;

    @DataUnit.AsOutput(name = "rdfOutput")
    public WritableRDFDataUnit output;

protected void innerExecute() throws DPUException {
   final List<RDFDataUnit.Entry> source =  DataUnitUtils.getMetadataEntries(input);
   final RDFDataUnit.Entry target =  DataUnitUtils.getWritableMetadataEntry(output);
   final SparqlUtils.SparqlUpdateObject update = SparqlUtils.createInsert(QUERY_COPY, source, target);
   // Copy data.
   SparqlUtils.execute(connection, update);
}
```
what is @tomas-knap @skodape opinion - does this merge all input graphs (symbolic names) into one outputgraph (symbolic name) or does this simply moves input symbolic names 1:1 to output symbolic names?

So when youhave 2 graphs on input
1. does this produce one graph on output containing triples from both input graphs 
2. or does this produce two output graphs with copied triples respectively

**Comment by mkl-eea-sk** [comment]
From the code above, I think, it should produce one graph on output. However, @jindrichmynarz reports it produces two.

In 1.x rdfMerger (core) it was producing two graphs, but that is not answer to bug, that I think must be in this code. Semantics seems to be 'produce one graph'. Something is wrong. @skodape : what was the intention? create one graph or two graphs? if one graph (ok, lets forget its a change of dpu behavior between 1.x and 2.x), then how this miraculously can produce two graphs for @jindrichmynarz ... then it is a bug.

**Comment by skodape** [comment]
The important note here is utilization of get__Metadata__Entries.

The intention was to create two graphs - like in the previous version (1.x).

**Comment by jindrichmynarz** [comment]
@tomas-knap: What's the use case for merging RDFDataUnits into 1 RDFDataUnit without merging data? I can't see how this DPU could be useful without merging actual data. Moreover, I believe users would expect a DPU named "RDF merger" to merge RDF graphs and not (some obscure internal concept of) RDFDataUnits.

**Comment by mkl-eea-sk** [comment]
@skodape hm, but the "target" is one entry, from sparqlutils:
```
 public static SparqlUpdateObject createInsert(String query, List<RDFDataUnit.Entry> sources,
            RDFDataUnit.Entry target) throws SparqlProblemException, DataUnitException {
        if (!useDataset()) {
            query = query.replaceFirst("(?i)INSERT", prepareClause("WITH", target) + "INSERT");
            query = query.replaceFirst("(?i)WHERE", prepareClause("USING", sources) + "WHERE");
        }
        // Return new object.
        return new SparqlUpdateObject(query, prepareDataset(sources, target));
    }
```
using and with - does this indicate creating two graphs. ok. i am not a sparql expert. but the code is, i dont know what to say. does this seems strange only to me?

**Comment by mkl-eea-sk** [comment]
It produces sparql:
```
with one_target_graph insert ... using source1, source2, source3 where ...
```
how this can produce two graphs...

**Comment by mkl-eea-sk** [comment]
from sparqlutils
```
   protected static Dataset prepareDataset(List<RDFDataUnit.Entry> source, RDFDataUnit.Entry target)
            throws DataUnitException {
        final DatasetImpl dataset = new DatasetImpl();
        // Add read graphs.
        for (URI uri : RdfDataUnitUtils.asGraphs(source)) {
            dataset.addDefaultGraph(uri);
        }
        // Add write graph.
        if (target != null) {
            dataset.setDefaultInsertGraph(target.getDataGraphURI());
            dataset.addDefaultRemoveGraph(target.getDataGraphURI());
        }
        return dataset;
    }
```
even dataset is prepared in a way - one target insert and delete graph, more source default graphs. 
@skodape either you must explain thoroughly or fix.

**Comment by mkl-eea-sk** [comment]
@jindrichmynarz please provide minimal pipeline which triggers error, i will try to reproduce and debug rdfMerger

**Comment by skodape** [comment]
@mkl-eea-sk Yes target is one entry, but you are working on *metadata* level, so you are effectively coping metadata = name of graphs that are in given DataUnit, not data it self.

**Comment by mkl-eea-sk** [comment]
@skodape THANKS! got the point

**Comment by jindrichmynarz** [comment]
@mkl-eea-sk: Please see [this pipeline](https://dl.dropboxusercontent.com/u/893551/pracovni/EEA/COMSODE/Pipelines/Test_t-rdfMerger.zip). Since you currently can't distinguish named graphs in UnifiedViews by querying the debug execution data (UnifiedViews/Core#450), this pipeline loads the merged RDFDataUnits into Virtuoso via l-rdfToVirtuoso. After you run this pipeline, go to Virtuoso SPARQL endpoint and retrieve list of named graphs:

```
SELECT DISTINCT ?g
WHERE {
  GRAPH ?g {
    [] ?p [] .
  }
}
```

The returned bindings should include 2 named graphs created by the pipeline (`http://unifiedviews.eu/resource/internal/dataunit/exec/506/dpu/620/du/1/user//0` and `http://unifiedviews.eu/resource/internal/dataunit/exec/506/dpu/620/du/1/user//1` in my case).

Tangential note: Why is there double slash in the auto-generated named graph URIs, such as `//1` in `http://unifiedviews.eu/resource/internal/dataunit/exec/506/dpu/620/du/1/user//1`?

**Comment by skrchnavy** [comment]
@tomas-knap / @jindrichmynarz is this DPU used in many pipelines? Was it necessary to use in current implementation instead of just connect few DPUs into one data unit of a *sparql* DPU?
*(Intention: if it is safe to put it into 2.1.1 release or have to wait for 2.2.0)*

**Comment by jindrichmynarz** [comment]
In the pipelines I have been working on this DPU is used several times. Typically, it is used when a pipeline links external data by extracting them via e-sparqlEndpoint, linking via t-sparqlConstruct, and then merging the links into the processed dataset via t-rdfMerger. However, I haven't found a use case for t-rdfMerger in its current form, in which it merges only RDFDataUnits and not data itself. The typical use case I have described would work only if the DPU could merge data.

Currently, there are work-arounds to be able to merge RDF data:

* Use t-sparqlConstruct without per graph mode with `CONSTRUCT WHERE {?s ?p ?o}` that just copies the data into a single graph.
* (And an uglier one:) Use t-rdfToFiles that implicitly merges named graphs (side note: why is the "Merge graphs" even in the DPU's dialog if the user cannot untick it?) followed by t-filesToRdf to get the data back into an RDFDataUnit.

**Comment by mkl-eea-sk** [comment]
```
DELETE FROM <http://unifiedviews../dataunit/exec/83/dpu/36/du/1>
INSERT INTO <http://unifiedviews../dataunit/exec/83/dpu/36/du/1>
USING <http://unifiedviews../dataunit/exec/83/dpu/37/du/1>
USING <http://unifiedviews../dataunit/exec/83/dpu/36/du/0>
```

**Comment by mkl-eea-sk** [comment]
OK, i have investigated @skodape 's code. Well, Mr. Repackaging metadataGraphNames as Entry (one symbolic name)! Better than obfuscating code... It works on metadata level, so it produces two graphs.

**Comment by mkl-eea-sk** [comment]
@tomas-knap please give requirements, are we going to add combo/checkbox to rdfMerger? semantics : merge data into single graph vs. copy input graphs to output ?

metadata attached to graphs are lost when merging into single graph on output.

@jindrichmynarz what is wrong with t-sparqlConstruct you use? isn't it a solution for this issue? if so, why should we code same funcionality twice.

original rdfMerger was there to save time:
1. first when you design a pipeline and you have 10 DPUs outputs to be connected to two other DPUs - without rdfMerger, you have to draw 10 lines from each source DPU to first target DPU, and then 10 lines from 10 source DPUs to second target DPU. with rdfMerger you draw 10 lines to rdfMerger and then 2 lines from rdfMerger to both target DPUs
2. in runtime , where, the original rdfMerger (in Core) work in O(1) with respect to data and metadata triples count. This rdfMerger work in O(n) where n = metadata triples count. Still better than original concept (which worked in O(m) where m = data triples count , which is usually many times more than metadata triples count). so this rdfMerger is still quite fast and provides needed info for debugging (well, we could do if debug {} else {} to speed up even more - when pipeline is not in debug mode, nobody is interested in intermediate metadata triples).
So i would leave rdfMerger as it is (although code is very hard to understand and misleading...)

**Comment by mkl-eea-sk** [comment]
I have assigned this issue to @tomas-knap  to let him think of requirements (if any) and rephrase the description.

**Comment by jindrichmynarz** [comment]
@mkl-eea-sk: Your explanation why t-rdfMerger just merging RDFDataUnits is useful clarifies it for me. Thanks!

However, I think there is still an issue with the DPU's name and the expectations it leads to in users. I believe users would typically expect a DPU with such name to merge data. Instead, what t-rdfMerger does is quite generic. I don't even understand why we need to have both t-rdfMerger and t-filesMerger. Why not just have a generic merge DPU notwithstanding the type of data units being merged (as long as they are all of the same type)?

Regarding your comment on why t-sparqlConstruct cannot be used instead for the merge functionality, the answer is that it can, but I'm wondering about 2 minor issues:

1. It is not obvious that t-sparqlConstruct can be used in this way. I think typical user would be lost on this.
2. Is t-sparqlConstruct efficient for merging data? I assume it effectively does something like copy by value, so I wonder if there are no better ways to do this (e.g., [`ADD`](http://www.w3.org/TR/sparql11-update/#add) smaller graphs to the largest one to save shuffling too many triples around).

To address (1) a DPU template with better name (e.g., "RDF merger" to keep things confusing) can be created. (2) was of no concern so far for COMSODE SK dataset because we usually merge smaller datasets.

In summary, I can live with using t-sparqlConstruct for merging data and we can close this issue unless there are other requirements to be addressed.

**Comment by mkl-eea-sk** [comment]
@jindrichmynarz ad generic merge (not bound to rdf/files) dataunit -> too complicated for implementation (changes in core). cost - benefit > 0. what would be output of such dpu? files data unit? rdf data unit? it could meta data unit (abstract of these two) - but how would you handle connecting it to DPU which expects files data unit...

1. agree, if there is a requirement to special type sparql construct (to merge graphs) we can handle it in two ways i think:
a. do it programatically, as a new DPU (i would propose "graphs merger")
b. provide data.sql script with ODN installation such that there will be, as default, after installation, DPU template called "Graphs merger" which will be child of "t-sparqlConstruct" with prefedined configuration template - the SPARQL construct you proposed as solution.
2. a. and .b from above are same in performance. i do not know about any more efficient way how to merge triples (this was the reason we removed code doing this type of op for each graph edge a year ago). and it depends on database - finding which graph is larger may also be time consuming. depends on how the underlying engine is programmed.

I would advise you, to postpone data merging to the latest (and most effective moment) in pipeline you can. The one which are effective are:
1. rdfToFiles - it does not make any difference in performance if you export data from single merged graph or more unmerged graphs - so using this DPU in global mode (!perGraph) is effectively merging graphs
2. rdfToVirtuoso - loading from more than one input graph is even faster than loading from single large graph - as multithreading gets involved and single unit of work == one input graph
3. filesToVirtuoso - loading from more than one file is even faster as virtuoso use more than one thread to load the data (clustering is done per file). so avoid merging data into single large file before running filesToVirtuoso.
4. t-sparql* dpus may be executed against set of graphs on input - same performance as execution against one large graph.

**Comment by jindrichmynarz** [comment]
Regarding packaging the graph merging functionality either as DPU or template I don't have a preference. I can see advantages to both approaches. Separate DPU could get rid of the configuration by SPARQL, while DPU template could be more light-weight solution.

As you suggest, I use merging at the furthest possible point in the pipeline. Regarding the options you mention:

Ad 1: t-rdfToFiles does have a configuration option "Merge graphs" that is enable by default and cannot be changed by the user (i.e. `input` is `disabled`). What's the point of having such option in the configuration if the user cannot change it?

Ad 2 and 3: Typically, we deal with datasets in non-incremental ways. Therefore, each pipeline run requires the previous version of the data to be cleared and replaced with the new version. However, if we load multiple unmerged graphs into the same target graph, while the option "Clear destination graph before loading" is enabled, wouldn't we end up with only the last graph being in the target graph (since we clear the destination graph for each previous graph)? Do I understand it correctly?

**Comment by mkl-eea-sk** [comment]
2.3.
- A. with clear graph and target graph set in vaadin dialog (!perGraph mode) - you will end up exactly with what you want - clearing is performed on target graph (from vaadin dialog), loading is done to target graph after clearing, loading from more graphs into single - works. does you tests result in something else?
clea
- B. with clear graph and target graph == null in vaadin dialog (perGraph mode) - each target graph is cleared - one per one symbolic name. the targetGraph is 'virtualGraph' metadata (which we had discussion elsewhere on - where to set it).

**Comment by jindrichmynarz** [comment]
I have tested it and it works as you describe. However, the same is not true for l-filesToParliament. I will create a separate issue for that, because it is off-topic.

I believe we can now close this issue and mark it as invalid, since I was asking for a functionality that is outside of the scope of t-rdfMerger. Correct?

**Comment by tomas-knap** [comment]
Summary. Lets close this issue, as physical merge will be implemented by a different DPU, see: https://github.com/UnifiedViews/Plugins/issues/272

I discuss that with CUNI, extra DPU is better for them, it makes sense for them to sometime physically merge and sometimes not. From what @jindrichmynarz said, extra DPU is ok for him. It could be also realized as a special case of t-sparqlConstruct, but I suppose that having new DPU is less confusing for the user and also easier to be automatically deployed to UV/maintened/further extended. 

Note: Ad t-rdfToFiles does have a configuration option "Merge graphs" that is enable by default and cannot be changed by the user (i.e. input is disabled). This is already removed in current develop branch, so use does not think about unchangable config option.

**Comment by skrchnavy** [comment]
What is resolution of this task. Invalid, fixed, wan't fix, ... ? Milestone have to be updated accordingly

**Comment by jindrichmynarz** [comment]
I'd say `invalid` or `wontfix`. I reported the issue as a bug of t-rdfMerger, but the DPU isn't intended to work the way I supposed it should.
